starknet-replay/src/storage/rpc/rpc_client.rs

```diff
@@ -0,0 +1,429 @@
+    /// [`RpcClient::starknet_get_chain_id`]. This is doable because it's not
+    /// possible to replay blocks from different chains.
+    chain_id: OnceCell<ChainId>,
+
+    /// The most recent block number available from the rpc server. It is
+    /// initialised in [`RpcClient::starknet_block_number`]. The assumption
+    /// is that the latest blocks generated after the replay is started are
+    /// ignored.
+    block_number: OnceCell<BlockNumber>,
+}
+impl RpcClient {
+    /// Constructs a new `RpcStorage`.
+    ///
+    /// # Arguments
+    ///
+    /// - `endpoint`: The Url of the Starknet RPC node.
+    #[must_use]
+    pub fn new(endpoint: Url) -> Self {
+        RpcClient {
+            endpoint,
+            chain_id: OnceCell::new(),
+            block_number: OnceCell::new(),
+        }
+    }
+
+    /// This function generates a new client to perform an RPC request to the
+    /// endpoint.
+    ///
+    /// The client can't be shared across threads.
+    fn get_new_client(&self) -> JsonRpcClient<HttpTransport> {
+        JsonRpcClient::new(HttpTransport::new(self.endpoint.clone()))
+    }
+
+    /// This function queries the number of the most recent Starknet block.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Err`] if the request fails.
+    #[allow(clippy::missing_panics_doc)] // Needed because `tokio::main` calls `unwrap()`
+    #[tokio::main]
+    pub async fn starknet_block_number(&self) -> Result<BlockNumber, DatabaseError> {
+        let block_number: Result<&BlockNumber, DatabaseError> = self
+            .block_number
+            .get_or_try_init(|| async {
+                let block_number: u64 = self.get_new_client().block_number().await?;
+                Ok(BlockNumber::new(block_number))
+            })
+            .await;
+        Ok(block_number?).cloned()
+    }
+
+    /// This function queries the contract class at a specific block.
+    ///
+    /// # Arguments
+    ///
+    /// - `class_hash_at_block`: class hash of the contract to be returned.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Err`] if the request fails or the class hash doesn't exist.
+    #[allow(clippy::missing_panics_doc)] // Needed because `tokio::main` calls `unwrap()`
+    #[tokio::main]
+    pub async fn starknet_get_class(
+        &self,
+        class_hash_at_block: &ReplayClassHash,
+    ) -> Result<ContractClass, DatabaseError> {
+        let block_id: BlockId = class_hash_at_block.block_number.into();
+        let class_hash: Felt = class_hash_at_block.class_hash.0;
+        let contract_class: ContractClass = self
+            .get_new_client()
+            .get_class(block_id, class_hash)
+            .await?;
+        Ok(contract_class)
+    }
+
+    /// This function queries the block header.
+    ///
+    /// # Arguments
+    ///
+    /// - `block_number`: the block number to be queried.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Err`] if the request fails or the block number doesn't exist.
```

Comment:
Same comment as above.

Reply:
This method is complex, it can return different types of error.